File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:

        if not s:
            return False

        p_map = {
            "(" : ")",
            "{" : "}",
            "[" : "]"
        }

        stack = []

        for ch in s:
            if ch in p_map:
                stack.append(ch)
            elif len(stack) == 0 or p_map[stack.pop()] != ch:
                return False

        return len(stack) == 0

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {

        Map<String, String> map = new HashMap<>();

        map.put("(", ")");
        map.put("{", "}");
        map.put("[", "]");

        List<String> stack = new ArrayList<>();

        for (int i = 0; i < s.length(); i++ ){

            char c = s.charAt(i);
            String str = String.valueOf(c);

            if (map.containsKey(str)) {
                stack.add(str);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }

                String lastOpen = stack.remove(stack.size() - 1);
                if (!map.get(lastOpen).equals(str)){
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
}